### DISCUSSION TITLE
Bind group validation errors in create_render_pipeline

### DISCUSSION THREAD

**Discussion by dhardy** [gfx-rs/wgpu#1432] [Q&A] [answered]
So I'm adjusting binding groups, attempting to divide a larger one into two. I've actually got things working now but would still like to know: what does this error mean?
```
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `images render pipeline`
    error in stage VERTEX
    error matching global binding at index 0 in group 0 against the pipeline layout: binding is missing from the pipeline layout
```
This occurred by removing all bind groups from the layout descriptor. I believe that is legal?

I also saw the error below (despite all `visibility` descriptors being set to either `FRAGMENT` or `VERTEX`), and another about texture dimension mis-match.
```
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `images render pipeline`
    error in stage VERTEX
    error matching global binding at index 0 in group 0 against the pipeline layout: visibility flags don't include the shader stage
```

I *guess* the errors might have been somehow due to incorrect incremental builds (I have had issues with those recently). In fact, I'm not sure what exactly solved the errors in the end. Just curious if you guys have any ideas why this might occur.

**Comment by kvark** [comment, marked as answer]
This error message is perhaps one of the most detailed ones we got recently. Since it's unclear still, we have a lot of room for improvement :) Perhaps, you could suggest the wording, assuming we can clarify what this means.

> error in stage VERTEX
> error matching global binding at index 0 in group 0 against the pipeline layout: binding is missing from the pipeline layout

It says that your vertex shader uses something with group=0 and binding=0, but the pipeline layout doesn't provide it. It can happen if the pipeline layout doesn't have any bind group layouts (BGLs), or the BGL[0] doesn't have binding=0, or there is a binding but its visibility doesn't include VERTEX.

The second posted error is the same as the first. Did you copy-paste by mistake?

**Reply by dhardy** [reply]
Aha, it wasn't clear to me that this is checking bind groups against *shader* requirements.

The second error is different (only the last bit about visibility). So from what you said I guess this error means that the vertex shader expects a binding at 0, 0 but this bind group is only available to the fragment shader?

Hmm, perhaps instead of `error in stage VERTEX` it could say `error evaluating requirements of vertex stage shader`?

**Reply by kvark** [reply]
Should be fixed in https://github.com/gfx-rs/wgpu/pull/1342
